### DISCUSSION TITLE
「重学TS 2.0 」TS 面试题：列举你工作中常用的 TS 内置工具类型并描述一下它们的作用

### DISCUSSION THREAD

**Discussion by semlinker** [semlinker/awesome-typescript#80] [General]
### 列举你工作中常用的 TS 内置工具类型并描述一下它们的作用

**Comment by wenkm** [comment]
`Partial` 转可选
`Required` 转必选
`Readonly` 转只读
`Exclude` 排除
`Extract` 提取
`Omit` 提取组成新的
`NonNullable` 排除null、undefined

**Comment by atom-7** [comment]
`ReturnType` 获取函数返回值类型

**Comment by douhao1988** [comment]
`Pick<T, K extends keyof T>`：从 T 类型中提取指定的成员组成一个新的类型

**Comment by pingan8787** [comment]
业务中常用的比较多：
- `Partial<Type>`：将指定类型的所有属性转换为可选类型，并且返回转换后的新类型；
- `Pick<Type, Keys>`：从指定类型中，选择一组需要的类型，并且返回选择后的新类型。常常在定义子类型的时候会去取。

完整的内置工具类型，可以查看官方文档学习：[Utility Types](https://www.typescriptlang.org/docs/handbook/utility-types.html)
